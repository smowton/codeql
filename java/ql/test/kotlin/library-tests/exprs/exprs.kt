import java.awt.Polygon
import java.awt.Rectangle

fun topLevelMethod(x: Int, y: Int): Int {
    val i1 = 1
    val i2 = x + y
    val i3 = x - y
    val i4 = x / y
    val i5 = x % y
/*
TODO
    val i6 = x shl y
    val i7 = x shr y
    val i8 = x ushr y
    val i9 = x and y
    val i10 = x or y
    val i11 = x xor y
    val i12 = x.inv()
*/
    val i13 = x == y
    val i14 = x != y
    val i15 = x < y
    val i16 = x <= y
    val i17 = x > y
    val i18 = x >= y
    val i19 = x === y
    val i20 = x !== y
/*
TODO
    val i20 = x in x .. y
    val i21 = x !in x .. y
*/
    val b1 = true
    val b2 = false
/*
TODO
    val b3 = b1 && b2
    val b4 = b1 || b2
    val b5 = !b1
*/
    val c = 'x'
    val str = "string lit"
    val strWithQuote = "string \" lit"
    val b6 = i1 is Int
    val b7 = i1 !is Int
    val b8 = b7 as Boolean
    val str1: String = "string lit"
    val str2: String? = "string lit"
    val str3: String? = null
    val str4: String = "foo $str1 bar $str2 baz"
    // TODO val str5: String = "foo ${str1 + str2} bar ${str2 + str1} baz"
    val str6 = str1 + str2

    var variable = 10
    while (variable > 0) {
        variable--
    }

    return 123 + 456
}

fun getClass() {
    val d = true::class
}

class C(val n: Int) {
    fun foo(): C { return C(42) }
}

open class Root {}
class Subclass1: Root() {}
class Subclass2: Root() {}

fun typeTests(x: Root, y: Subclass1) {
    if(x is Subclass1) {
        val x1: Subclass1 = x
    }
    val y1: Subclass1 = if (x is Subclass1) { x } else { y }
    var q = 1
    if (x is Subclass1) { q = 2 } else { q = 3 }
}

fun foo(p: Polygon) {
    val r = p.getBounds()
    if(r != null) {
        val r2: Rectangle = r
        val height = r2.height
        r2.height = 3
    }
}

enum class Direction {
    NORTH, SOUTH, WEST, EAST
}

enum class Color(val rgb: Int) {
    RED(0xFF0000),
    GREEN(0x00FF00),
    BLUE(0x0000FF)
}

fun enums() {
    val south = Direction.SOUTH
    val green = Color.GREEN
}
