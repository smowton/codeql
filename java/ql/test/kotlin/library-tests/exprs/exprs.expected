| exprs.kt:5:5:5:14 | i1 | LocalVariableDeclExpr |
| exprs.kt:5:14:5:14 | 1 | IntegerLiteral |
| exprs.kt:6:5:6:18 | i2 | LocalVariableDeclExpr |
| exprs.kt:6:14:6:14 | x | VarAccess |
| exprs.kt:6:14:6:18 | ... + ... | AddExpr |
| exprs.kt:6:18:6:18 | y | VarAccess |
| exprs.kt:7:5:7:18 | i3 | LocalVariableDeclExpr |
| exprs.kt:7:14:7:14 | x | VarAccess |
| exprs.kt:7:14:7:18 | ... - ... | SubExpr |
| exprs.kt:7:18:7:18 | y | VarAccess |
| exprs.kt:8:5:8:18 | i4 | LocalVariableDeclExpr |
| exprs.kt:8:14:8:14 | x | VarAccess |
| exprs.kt:8:14:8:18 | ... / ... | DivExpr |
| exprs.kt:8:18:8:18 | y | VarAccess |
| exprs.kt:9:5:9:18 | i5 | LocalVariableDeclExpr |
| exprs.kt:9:14:9:14 | x | VarAccess |
| exprs.kt:9:14:9:18 | ... % ... | RemExpr |
| exprs.kt:9:18:9:18 | y | VarAccess |
| exprs.kt:20:5:20:20 | i13 | LocalVariableDeclExpr |
| exprs.kt:20:15:20:15 | x | VarAccess |
| exprs.kt:20:15:20:20 | ... == ... | EQExpr |
| exprs.kt:20:20:20:20 | y | VarAccess |
| exprs.kt:21:5:21:20 | i14 | LocalVariableDeclExpr |
| exprs.kt:21:15:21:15 | x | VarAccess |
| exprs.kt:21:15:21:20 | ... != ... | NEExpr |
| exprs.kt:21:15:21:20 | ... != ... | NEExpr |
| exprs.kt:21:20:21:20 | y | VarAccess |
| exprs.kt:22:5:22:19 | i15 | LocalVariableDeclExpr |
| exprs.kt:22:15:22:15 | x | VarAccess |
| exprs.kt:22:15:22:19 | ... < ... | LTExpr |
| exprs.kt:22:19:22:19 | y | VarAccess |
| exprs.kt:23:5:23:20 | i16 | LocalVariableDeclExpr |
| exprs.kt:23:15:23:15 | x | VarAccess |
| exprs.kt:23:15:23:20 | ... <= ... | LEExpr |
| exprs.kt:23:20:23:20 | y | VarAccess |
| exprs.kt:24:5:24:19 | i17 | LocalVariableDeclExpr |
| exprs.kt:24:15:24:15 | x | VarAccess |
| exprs.kt:24:15:24:19 | ... > ... | GTExpr |
| exprs.kt:24:19:24:19 | y | VarAccess |
| exprs.kt:25:5:25:20 | i18 | LocalVariableDeclExpr |
| exprs.kt:25:15:25:15 | x | VarAccess |
| exprs.kt:25:15:25:20 | ... >= ... | GEExpr |
| exprs.kt:25:20:25:20 | y | VarAccess |
| exprs.kt:31:5:31:17 | b1 | LocalVariableDeclExpr |
| exprs.kt:31:14:31:17 | true | BooleanLiteral |
| exprs.kt:32:5:32:18 | b2 | LocalVariableDeclExpr |
| exprs.kt:32:14:32:18 | false | BooleanLiteral |
| exprs.kt:39:5:39:15 | c | LocalVariableDeclExpr |
| exprs.kt:39:13:39:15 | x | CharacterLiteral |
| exprs.kt:40:5:40:26 | str | LocalVariableDeclExpr |
| exprs.kt:40:16:40:25 | string lit | StringLiteral |
| exprs.kt:41:5:41:38 | strWithQuote | LocalVariableDeclExpr |
| exprs.kt:41:25:41:37 | string " lit | StringLiteral |
| exprs.kt:42:5:42:22 | b6 | LocalVariableDeclExpr |
| exprs.kt:42:14:42:15 | i1 | VarAccess |
| exprs.kt:42:14:42:22 | ...instanceof... | InstanceOfExpr |
| exprs.kt:42:14:42:22 | int | TypeAccess |
| exprs.kt:43:5:43:23 | b7 | LocalVariableDeclExpr |
| exprs.kt:43:14:43:15 | i1 | VarAccess |
| exprs.kt:43:14:43:23 | ... !is ... | NotInstanceOfExpr |
| exprs.kt:43:14:43:23 | int | TypeAccess |
| exprs.kt:44:5:44:26 | b8 | LocalVariableDeclExpr |
| exprs.kt:44:14:44:15 | b7 | VarAccess |
| exprs.kt:44:14:44:26 | (...)... | CastExpr |
| exprs.kt:44:14:44:26 | boolean | TypeAccess |
| exprs.kt:45:5:45:35 | str1 | LocalVariableDeclExpr |
| exprs.kt:45:25:45:34 | string lit | StringLiteral |
| exprs.kt:46:5:46:36 | str2 | LocalVariableDeclExpr |
| exprs.kt:46:26:46:35 | string lit | StringLiteral |
| exprs.kt:47:5:47:28 | str3 | LocalVariableDeclExpr |
| exprs.kt:47:25:47:28 | null | NullLiteral |
| exprs.kt:49:5:49:21 | variable | LocalVariableDeclExpr |
| exprs.kt:49:20:49:21 | 10 | IntegerLiteral |
| exprs.kt:50:12:50:19 | variable | VarAccess |
| exprs.kt:50:12:50:23 | ... > ... | GTExpr |
| exprs.kt:50:23:50:23 | 0 | IntegerLiteral |
| exprs.kt:51:9:51:16 | ...=... | AssignExpr |
| exprs.kt:51:9:51:16 | variable | VarAccess |
| exprs.kt:51:9:51:18 | (...)... | CastExpr |
| exprs.kt:51:9:51:18 | <Stmt> | StmtExpr |
| exprs.kt:51:9:51:18 | dec(...) | MethodAccess |
| exprs.kt:51:9:51:18 | tmp0 | VarAccess |
| exprs.kt:51:9:51:18 | tmp0 | VarAccess |
| exprs.kt:51:9:51:18 | void | TypeAccess |
| exprs.kt:54:12:54:14 | 123 | IntegerLiteral |
| exprs.kt:54:12:54:20 | ... + ... | AddExpr |
| exprs.kt:54:18:54:20 | 456 | IntegerLiteral |
| exprs.kt:58:5:58:23 | d | LocalVariableDeclExpr |
| exprs.kt:58:13:58:16 | true | BooleanLiteral |
| exprs.kt:58:13:58:23 | ::class | ClassExpr |
| exprs.kt:61:1:63:1 | <obinit>(...) | MethodAccess |
| exprs.kt:61:9:61:18 | ...=... | AssignExpr |
| exprs.kt:61:9:61:18 | n | VarAccess |
| exprs.kt:61:9:61:18 | n | VarAccess |
| exprs.kt:62:27:62:31 | new C(...) | ClassInstanceExpr |
| exprs.kt:62:29:62:30 | 42 | IntegerLiteral |
| exprs.kt:65:1:65:18 | <obinit>(...) | MethodAccess |
| exprs.kt:66:1:66:26 | <obinit>(...) | MethodAccess |
| exprs.kt:67:1:67:26 | <obinit>(...) | MethodAccess |
| exprs.kt:70:5:72:5 | when ... | WhenExpr |
| exprs.kt:70:8:70:8 | x | VarAccess |
| exprs.kt:70:8:70:21 | ...instanceof... | InstanceOfExpr |
| exprs.kt:70:8:70:21 | Subclass1 | TypeAccess |
| exprs.kt:71:9:71:28 | y | LocalVariableDeclExpr |
| exprs.kt:71:28:71:28 | (...)... | CastExpr |
| exprs.kt:71:28:71:28 | Subclass1 | TypeAccess |
| exprs.kt:71:28:71:28 | x | VarAccess |
| exprs.kt:76:5:76:25 | r | LocalVariableDeclExpr |
| exprs.kt:76:13:76:13 | p | VarAccess |
| exprs.kt:76:15:76:25 | getBounds(...) | MethodAccess |
| exprs.kt:77:5:79:5 | when ... | WhenExpr |
| exprs.kt:77:8:77:8 | r | VarAccess |
| exprs.kt:77:8:77:16 | ... != ... | NEExpr |
| exprs.kt:77:8:77:16 | ... != ... | NEExpr |
| exprs.kt:77:13:77:16 | null | NullLiteral |
| exprs.kt:78:9:78:29 | r2 | LocalVariableDeclExpr |
| exprs.kt:78:29:78:29 | (...)... | CastExpr |
| exprs.kt:78:29:78:29 | Rectangle | TypeAccess |
| exprs.kt:78:29:78:29 | r | VarAccess |
| file://:0:0:0:0 | tmp0 | LocalVariableDeclExpr |
| file://:0:0:0:0 | variable | VarAccess |
